### PR TITLE
feat: Support Panes for AppFlowy

### PR DESCRIPTION
**Pull request by squidrye**
### Feature Preview

https://github.com/AppFlowy-IO/AppFlowy/assets/84044317/152fddd2-3948-4bfa-b3b5-2fc90fde8bd5

---
#2312 This PR targets implementation of multi-panes
cc @Xazin 

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

#### Todos
- [x] Decide and implement data structure for panes
- [x] Map n-ary tree to bloc for state management, allow opening and closing of panes
- [x] Refactor existing tabs implementation to each pane node
- [x] Link each pane to correct plugin and view 
- [x] Resizable panes
- [ ] Setup drag and drop for tabs
- [ ] Setup drag and drop for panes
- [ ] Allow horizontal/ vertical scroll for stacked panes based on visible viewport
- [x] Stack panes over another (Sliding panes)
- [ ] Refactor other features which might break due to this (Mostly scroll related issues)
- [ ] Might need a better design spec for tab-bar, with/without tabs
- [ ] Tests

### CODE REVIEW

File: frontend/appflowy_flutter/lib/workspace/presentation/home/home_stack.dart [hunk 1]
```diff
@@ -253,9 +300,11 @@ class PageManager {
 }
 
 class HomeTopBar extends StatelessWidget {
-  const HomeTopBar({Key? key, required this.layout}) : super(key: key);
+  const HomeTopBar({Key? key, required this.layout, required this.paneId})
+      : super(key: key);
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/home_stack.dart:304]:
```suggestion
  const HomeTopBar({super.key, required this.layout, required this.paneId});
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/home_stack.dart [hunk 2]
```diff
@@ -280,6 +329,19 @@ class HomeTopBar extends StatelessWidget {
                     const SizedBox.shrink(),
               ),
             ),
+            BlocBuilder<PanesCubit, PanesState>(
+              builder: (context, state) {
+                Log.warn("Count ${state.count}");
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/home_stack.dart:334]: Remove

File: frontend/appflowy_flutter/lib/workspace/presentation/home/hotkeys.dart [hunk 3]
```diff
@@ -97,7 +97,10 @@ class HomeHotKeys extends StatelessWidget {
   }
 
   void _selectTab(BuildContext context, int change) {
-    final bloc = context.read<TabsBloc>();
-    bloc.add(TabsEvent.selectTab(bloc.state.currentIndex + change));
+    final bloc = context.read<PanesCubit>();
+    // bloc.add(TabsEvent.selectTab(bloc.state.currentIndex + change));
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/hotkeys.dart:101]: Remember to remove all comments and double check once you're ready for that.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/mention/mention_page_block.dart [hunk 4]
```diff
@@ -108,12 +108,16 @@ class _MentionPageBlockState extends State<MentionPageBlock> {
...
+    getIt<PanesCubit>().openTab(
+      plugin: view.plugin(),
+      view: view,
     );
+    // getIt<TabsBloc>().add(
+    //   TabsEvent.openPlugin(
+    //     plugin: view.plugin(),
+    //     view: view,
+    //   ),
+    // );
```
[the remark this answers is not in the document]
Comment (squidrye) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/mention/mention_page_block.dart:120]: Will do 👍🏼

File: frontend/appflowy_flutter/lib/workspace/application/panes/cubit/panes_cubit.dart [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import 'package:bloc/bloc.dart';
+import 'package:equatable/equatable.dart';
+import 'package:flutter/material.dart';
+
+import '../panes.dart';
+
+part 'panes_state.dart';
+
+class PanesCubit extends Cubit<PanesState> {
+  static final key = UniqueKey().toString();
```
[the remark this answers is not in the document]
Comment (squidrye) [reply frontend/appflowy_flutter/lib/workspace/application/panes/cubit/panes_cubit.dart:17]: I am not using it anymore, I will remove it, though it did its job, I just wanted to use it for unique comparisons of panes

File: frontend/appflowy_flutter/lib/workspace/application/panes/panes_service.dart [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  PaneNode closePaneHandler(
+    PaneNode node,
+    String targetPaneId,
+    Function(PaneNode) setActiveNode,
+  ) {
+    if (node.paneId == targetPaneId) {
+      return node;
+    } else {
```
[the remark this answers is not in the document]
Comment (squidrye) [reply frontend/appflowy_flutter/lib/workspace/application/panes/panes_service.dart:93]: Whoops, my bad

**Comment by squidrye** [comment]
I have updated tasks in the PR description, right now I am trying to figure out how to enable scrolling based on visible viewport of each pane, I don't want to constraint size of panes as that wraps the content and doesn't provide a good user experience, so I have stacked them (Sliding Panes) based on some calculations . 

Any help regarding how to enable scroll would be appreciated. Here's how it looks right now.


https://github.com/AppFlowy-IO/AppFlowy/assets/84044317/7c98c24b-a965-4505-afaa-a5b985a72ef1


ps: I'll do something about the blue borders later on.

**Comment by squidrye** [comment]
Opened a new PR, had to delete my older fork because I had pushed private code due to local git misconfiguration 😢

Apologies for inconvenience

**Pull request closed without merging**